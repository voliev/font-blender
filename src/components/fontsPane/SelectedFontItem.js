import React, { Component } from "react";
import RemoveIcon from "@material-ui/icons/RemoveCircleOutlineRounded";
import PropTypes from "prop-types";
import "./styles/SelectedFontItem.css";

class SelectedFontItem extends Component {
  constructor() {
    super();

    this.item = React.createRef();

    this.handleButtonClick = this.handleButtonClick.bind(this);
    this.handleTransitionEnd = this.handleTransitionEnd.bind(this);
  }

  handleButtonClick(e) {
    const { target } = e;

    if (target.closest("button")) {
      this.item.current.classList.remove("active");
    }
  }

  handleTransitionEnd(e) {
    e.persist();

    if (e.propertyName === "transform") {
      return;
    }

    const { target } = e;
    const { family, fontDeselectionHandler } = this.props;

    if (!target.classList.contains("active")) {
      fontDeselectionHandler(family);
    }
  }

  componentDidMount() {
    if (this.item.current) {
      this.item.current.classList.add("active");
    }
  }

  render() {
    const { family } = this.props;

    return (
      <li
        className="SelectedFontItem"
        onTransitionEnd={this.handleTransitionEnd}
        ref={this.item}
      >
        <p className="font-family">{family}</p>
        <button
          className="btn btn-round-small"
          onClick={this.handleButtonClick}
          title="Deselect font"
          value={family}
        >
          <RemoveIcon fontSize="default" />
        </button>
      </li>
    );
  }
}

SelectedFontItem.propTypes = {
  family: PropTypes.string.isRequired,
  fontDeselectionHandler: PropTypes.func.isRequired
};

export default SelectedFontItem;
