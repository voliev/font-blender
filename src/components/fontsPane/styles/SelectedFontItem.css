.SelectedFontItem {
  align-items: center;
  display: flex;
  font-weight: 700;
  justify-content: space-between;
  height: 0;
  opacity: 0;
  overflow: hidden;
  transition: height 0.2s, opacity 0.2s;

}

.SelectedFontItem.active {
  height: 2.75rem;
  opacity: 1;
  transition: height 0.2s, opacity 0.2s;
}

.SelectedFontItem:last-child {
  margin: 0;
}

.SelectedFontItem .font-family {
  margin: 0 0 0 var(--space-lg);
}

.SelectedFontItem .btn {
  color: var(--red);
  margin: 0 var(--space-lg) 0 0;
  transition: color 0.1s ease-out, transform 0.1s ease-out;
}

.SelectedFontItem .btn:hover {
  transform: scale(1.2);
}
