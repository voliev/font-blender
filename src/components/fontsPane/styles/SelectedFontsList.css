.SelectedFontsList {
  background-color: var(--white);
  list-style: none;
  margin: 0;
  overflow-y: auto;
  padding: var(--space-md) 0;
}
