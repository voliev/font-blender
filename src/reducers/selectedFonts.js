import undoable, { includeAction } from "redux-undo";
import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { getFonts } from "./fonts";

const initialState = [];

export const selectedFonts = (state = initialState, action) => {
  switch (action.type) {
    case ADD_SELECTED_FONT:
      return [...state, action.family];
    case REMOVE_FONT_FROM_SELECTED:
      return state.filter(family => family !== action.family);
    case CLEAR_SELECTED_FONTS:
      return [];
    default:
      return state;
  }
};

export default undoable(selectedFonts, {
  limit: 1,
  filter: includeAction([REMOVE_FONT_FROM_SELECTED, CLEAR_SELECTED_FONTS]),
  undoType: UNDO_DESELECTED
});

export const getSelectedFontFamilies = state => state.selectedFonts.present;
export const getSelectedFonts = state => {
  const fonts = getFonts(state);

  return getSelectedFontFamilies(state).reduce((selected, family) => {
    selected[family] = fonts[family];
    return selected;
  }, {});
};
export const isSelected = (state, family) =>
  state.selectedFonts.present.includes(family);
