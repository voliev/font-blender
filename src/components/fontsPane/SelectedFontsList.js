import React from "react";
import PropTypes from "prop-types";
import SelectedFontItem from "./SelectedFontItem";
import "./styles/SelectedFontsList.css";

const SelectedFontsList = ({ fontDeselectionHandler, selectedFonts }) => (
  <ul className="SelectedFontsList">
    {selectedFonts.map(family => (
      <SelectedFontItem
        family={family}
        key={family}
        fontDeselectionHandler={fontDeselectionHandler}
      />
    ))}
  </ul>
);

SelectedFontsList.propTypes = {
  fontDeselectionHandler: PropTypes.func.isRequired,
  selectedFonts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired
};

export default SelectedFontsList;
