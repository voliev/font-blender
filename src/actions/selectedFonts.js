import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { startFetchFont } from "./fonts";
import { addNotification, removeNotification } from "./notifications";
import { getFonts } from "../reducers/fonts";

export const addFontToSelected = family => ({
  type: ADD_SELECTED_FONT,
  family
});

export const removeFontFromSelected = family => ({
  type: REMOVE_FONT_FROM_SELECTED,
  family
});

export const clearSelectedFonts = () => ({
  type: CLEAR_SELECTED_FONTS
});

export const startAddFontToSelected = family => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];

    if (font.fetched) {
      // Add font to selected if it has been already fetchedselected list
      dispatch(addFontToSelected(family));
    } else {
      // Fetch font first than add it to selected
      return dispatch(startFetchFont(family))
        .then(() => dispatch(addFontToSelected(family)))
        .catch(() => {});
    }
  };
};

export const undoDeselected = () => ({
  type: UNDO_DESELECTED
});

export const startRemoveFontFromSelected = family => {
  return dispatch => {
    dispatch(removeFontFromSelected(family));

    const text = `${family} font family has been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startClearSelectedFonts = () => {
  return (dispatch, getState) => {
    const state = getState();
    const selectedFonts = [...state.selectedFonts.present];

    dispatch(clearSelectedFonts());

    const text =
      selectedFonts.length > 1
        ? `${selectedFonts.length} font families have been removed`
        : `${selectedFonts.length} font family have been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startUndoDeselected = () => {
  return dispatch => {
    // Undo deselection
    dispatch(undoDeselected());

    const group = "selectedFonts";

    // Remove notification to hide "undo" option
    dispatch(removeNotification(group));
  };
};
