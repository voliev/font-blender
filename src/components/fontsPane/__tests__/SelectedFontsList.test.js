import React from "react";
import { mount, shallow } from "enzyme";
import SelectedFontsList from "../SelectedFontsList";
import SelectedFontItem from "../SelectedFontItem";
import selectedFonts from "../../../fixtures/selectedFonts";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const props = {
    fontDeselectionHandler: jest.fn(),
    selectedFonts,
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<SelectedFontsList {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const { wrapper, selectedFonts, fontDeselectionHandler } = setup(mount);

  expect(typeof wrapper.prop("selectedFonts")).toBe("object");
  expect(wrapper.prop("selectedFonts").hasOwnProperty("length")).toBeTruthy();
  expect(wrapper.prop("selectedFonts")).toEqual(selectedFonts);

  expect(typeof wrapper.prop("fontDeselectionHandler")).toBe("function");
  expect(wrapper.prop("fontDeselectionHandler")).toEqual(
    fontDeselectionHandler
  );
});

it("should render list items", () => {
  const { selectedFonts, wrapper } = setup();

  expect(wrapper.find(SelectedFontItem)).toHaveLength(selectedFonts.length);
});
