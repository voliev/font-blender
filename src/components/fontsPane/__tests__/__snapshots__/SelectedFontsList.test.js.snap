// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<ul
  className="SelectedFontsList"
>
  <SelectedFontItem
    family="Lato"
    fontDeselectionHandler={[MockFunction]}
    key="Lato"
  />
  <SelectedFontItem
    family="Montserrat"
    fontDeselectionHandler={[MockFunction]}
    key="Montserrat"
  />
  <SelectedFontItem
    family="Roboto Condensed"
    fontDeselectionHandler={[MockFunction]}
    key="Roboto Condensed"
  />
  <SelectedFontItem
    family="Raleway"
    fontDeselectionHandler={[MockFunction]}
    key="Raleway"
  />
</ul>
`;
